### PR TITLE
Add QgsProject::dirtySet signal

### PR DESCRIPTION
**Pull request by manisandro**
Adds a QgsProject::dirtySet signal, emitted when QgsProject::setDirty(true) is invoked.

### CODE REVIEW

**Comment by m-kuhn** [comment]
Isn't that redundant with `isDirtyChanged( bool )`?

**Comment by manisandro** [comment]
I believe `isDirtyChanged` doesn't allow you to distinguish whether `setDirty` was called with `true` or `false` as parameter. in KADAS I use `dirtySet` to trigger an autosave, but I don't want to trigger an autosave when the project was just saved (hence `setDirty(false)` is called.

**Comment by m-kuhn** [comment]
There is a bool emitted along with the signal (and you can always call `isDirty()` to check the status, is that not enough?

**Comment by manisandro** [comment]
Hmm, don't remember exactly the issue I had, but I'll check again.

**Comment by manisandro** [comment]
Ahh now I remember: the problem is that isDirtyChanged is only triggered when the dirty state changes. But I need to know every time `setDirty(true)` is called to trigger an autosave, even if the project state is already dirty.

**Comment by m-kuhn** [comment]
Wouldn't it be appropriate to set dirty = false after an autosave? Or does it save to a separate file?

**Comment by manisandro** [comment]
A separate file is created, more like an automatic backup. When opening a project file, if a more recent backup file exists, the user is asked whether to open that one rather than the project. But the project file itself will only be written when the user explicitly saves.

File: src/core/project/qgsproject.h [hunk 1]
```diff
@@ -1824,6 +1824,13 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     void isDirtyChanged( bool dirty );
 
+    /**
+     * Emitted when QgsProject::setDirty(true) is called.
```
Comment (m-kuhn) [inline src/core/project/qgsproject.h:1828]: Can you add a description of the difference to isDirtyChanged

**Review by m-kuhn — APPROVED** [review]
Looks good to me, I think a note in the docs to highlight the differences would be good

File: python/core/auto_generated/project/qgsproject.sip.in [hunk 2]
```diff
@@ -1799,6 +1799,18 @@ Emitted when the project dirty status changes.
...
+    void dirtySet();
+%Docstring
+Emitted when setDirty(true) is called.
+
+.. note::
+
+   As opposed to :py:func:`~QgsProject.isDirtyChanged`, this signal is invoked every time setDirty(true)
+   is called, regardless of whether the project was already dirty.
+
+.. versionadded:: 3.22
```
Comment (nyalldawson) [inline python/core/auto_generated/project/qgsproject.sip.in:1811]:
```suggestion
.. versionadded:: 3.20
```

File: src/core/project/qgsproject.h [hunk 3]
```diff
@@ -1824,6 +1824,15 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     void isDirtyChanged( bool dirty );
 
+    /**
+     * Emitted when setDirty(true) is called.
+     * \note As opposed to isDirtyChanged(), this signal is invoked every time setDirty(true)
+     * is called, regardless of whether the project was already dirty.
+     *
+     * \since QGIS 3.22
```
Comment (nyalldawson) [inline src/core/project/qgsproject.h:1832]:
```suggestion
     * \since QGIS 3.20
```

**Comment by nyalldawson** [comment]
@manisandro seems I can't commit suggestions to your PRs. Can you update the since version and then we can merge this for 3.20.

**Pull request merged**